backup-stores/gcs/src/main/java/io/camunda/zeebe/backup/gcs/GcsBackupStore.java

```diff
@@ -7,34 +7,122 @@
+import io.camunda.zeebe.backup.gcs.GcsBackupStoreException.UnexpectedManifestState;
+import io.camunda.zeebe.backup.gcs.manifest.Manifest;
+import java.io.IOException;
 import java.nio.file.Path;
 import java.util.Collection;
+import java.util.Optional;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
 
 public final class GcsBackupStore implements BackupStore {
+  public static final ObjectMapper MAPPER =
+      new ObjectMapper()
+          .registerModule(new Jdk8Module())
+          .registerModule(new JavaTimeModule())
+          .disable(WRITE_DATES_AS_TIMESTAMPS)
+          .setSerializationInclusion(Include.NON_ABSENT);
+
   private final GcsBackupConfig config;
+  private final Storage client;
+  private final BucketInfo bucketInfo;
+  private final Executor executor;
 
   public GcsBackupStore(final GcsBackupConfig config) {
+    this(config, buildClient(config));
+  }
+
+  public GcsBackupStore(final GcsBackupConfig config, final Storage client) {
     this.config = config;
+    this.client = client;
+    bucketInfo = BucketInfo.of(config.bucketName());
+    executor = Executors.newWorkStealingPool(4);
   }
 
   @Override
   public CompletableFuture<Void> save(final Backup backup) {
-    throw new UnsupportedOperationException();
+    return CompletableFuture.runAsync(() -> saveSync(backup), executor);
+  }
+
+  private void saveSync(final Backup backup) {
+    final var inProgress = Manifest.create(backup);
+    final var blobInfo = manifestBlobInfo(backup.id());
+    final Blob initial;
+
+    try {
+
+      initial =
+          client.create(
+              blobInfo, MAPPER.writeValueAsBytes(inProgress), BlobTargetOption.doesNotExist());
+    } catch (final StorageException e) {
+      throw new UnexpectedManifestState(
+          "Manifest for backup %s already exists".formatted(backup.id()));
+    } catch (JsonProcessingException e) {
+      throw new RuntimeException(e);
+    }
+
+    final var completed = inProgress.complete();
+    try {
+      client.create(
+          blobInfo,
+          MAPPER.writeValueAsBytes(completed),
+          BlobTargetOption.generationMatch(initial.getGeneration()));
+    } catch (final StorageException e) {
+      throw new UnexpectedManifestState(
+          "Manifest for backup %s was modified unexpectedly".formatted(backup.id()));
+    } catch (JsonProcessingException e) {
+      throw new RuntimeException(e);
+    }
   }
 
   @Override
   public CompletableFuture<BackupStatus> getStatus(final BackupIdentifier id) {
-    throw new UnsupportedOperationException();
+    return CompletableFuture.supplyAsync(() -> getStatusSync(id));
+  }
+
+  private BackupStatus getStatusSync(final BackupIdentifier id) {
+    final var blob = client.get(manifestBlobInfo(id).getBlobId());
+    if (blob == null) {
+      return new BackupStatusImpl(
```

Comment:
:wrench: we could create a factor method
```suggestion
      return BackupStatusImpl.doesNotExist()
```